src/generators/mappersGenerator.ts

```diff
@@ -21,29 +28,132 @@ export function generateMappers(
+          );
+        }
+
+        if (uberParent === className) {
+          discriminators[className] = className;
+        } else {
+          discriminators[`${uberParent}.${mapper.serializedName}`] = className;
+        }
+      }
+
       mappersFile.addVariableStatement({
         isExported: true,
         declarations: [
           {
-            name: normalizeName(mapper.serializedName, NameType.Class),
+            name: normalizeName(
+              compositeMapper.type.className || "",
+              NameType.Class
+            ),
             type: "coreHttp.CompositeMapper",
-            initializer: mapperString
+            initializer: writer => writeMapper(writer, mapper)
           }
         ],
         declarationKind: VariableDeclarationKind.Const,
         leadingTrivia: writer => writer.blankLine()
       });
     } else {
-      throw new Error("Mapper does not have a className");
+      throw new Error(
+        `Don't know how to create a mapper for type ${mapper.type.name}`
+      );
     }
   }
+
+  !isEmpty(discriminators) &&
+    mappersFile.addVariableStatement({
+      isExported: true,
+      declarations: [
+        {
+          name: "discriminators",
+          initializer: writer => {
+            writer.block(() => {
+              keys(discriminators).forEach(key => {
+                writer.write(`'${key}': ${discriminators[key]},`);
+              });
+            });
+          }
+        }
+      ]
+    });
+}
+
+function writeMapper(writer: CodeBlockWriter, mapper: Mapper) {
+  const parents = extractParents(mapper);
+  const { type, ...restMapper } = mapper;
+  const {
+    modelProperties,
+    polymorphicDiscriminator,
+    ...restType
+  } = type as CompositeMapperType;
+  const isReferenceDicriminator = isString(polymorphicDiscriminator);
+  writer.block(() => {
+    writeObjectProps(restMapper, writer)
+      .write("type:")
+      .block(() => {
+        writeObjectProps(restType, writer)
+          .conditionalWrite(
+            !!polymorphicDiscriminator,
+            "polymorphicDiscriminator:"
+          )
+          .conditionalWrite(
+            polymorphicDiscriminator && isReferenceDicriminator,
+            `${polymorphicDiscriminator as any},`
+          );
+        !isReferenceDicriminator &&
+          polymorphicDiscriminator &&
+          writer
+            .block(() => {
+              writeObjectProps(polymorphicDiscriminator, writer);
+            })
+            .write(",");
+        writer.write("modelProperties:").block(() => {
+          writeParentMappers(parents, writer);
+          writeObjectProps(modelProperties, writer);
+        });
+      });
+  });
+}
+
+function writeParentMappers(parents: string[], writer: CodeBlockWriter) {
+  (parents || []).forEach(parent => {
+    writer.write(`...${parent}.type.modelProperties,`);
+  });
+}
+
+function writeObjectProps(obj: any, writer: CodeBlockWriter) {
+  let currentPos = writer;
+  keys(obj).map(key => {
```

Comment:
Technically a `forEach` is more appropriate in this case even though JavaScript's `map` evaluates immediately too
```suggestion
  keys(obj).forEach(key => {
```